optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,154 @@
+    Note that this implementation does not take as inputs the maximum amount of resource to
+    a single SHA noted as :math:`R` in the paper.
+
+    Args:
+        min_resource:
+            A parameter for specifying the minimum resource allocated to a trial noted as :math:`r`
+            in the paper.
+            See the details for :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        reduction_factor:
+            A parameter for specifying reduction factor of promotable trials noted as
+            :math:`\\eta` in the paper. See the details for
+            :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        n_brackets:
+            The number of :class:`~optuna.pruners.SuccessiveHalvingPruner`\\s (brackets).
+        min_early_stopping_rate_low:
+            A parameter for specifying the minimum early-stopping rate.
+            This parameter is related to a parameter that is referred to as :math:`r` and used in
+            `Asynchronous SuccessiveHalving paper <http://arxiv.org/abs/1810.05934>`_.
+            The minimum early stopping rate for ``i``th bracket is :math:`i + s`.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            n_brackets=4,
+            min_early_stopping_rate_low=0
+    ):
+        # type: (int, int, int, int) -> None
+
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_budget = 0
+        self._n_brackets = n_brackets
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hyperband has {} brackets'.format(self._n_brackets))
+
+        for i in range(n_brackets):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_brackets)
+            self._resource_budget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> bool
+
+        i = self._get_bracket_id(study, trial)
+        _logger.debug('{}th bracket is selected'.format(i))
+        bracket_study = self._create_bracket_study(study, i)
+        return self._pruners[i].prune(bracket_study, trial)
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_brackets):
+        # type: (int, int) -> int
```

Comment:
Add a blank line after type hints.

```suggestion
        # type: (int, int) -> int

```